worlds/witness/hints.py

```diff
@@ -338,105 +392,323 @@ def make_hints(world: "WitnessWorld", hint_amount: int, own_itempool: List[Item]
+    unhinted_locations_per_area = dict()
+    unhinted_location_percentage_per_area = dict()
 
-        already_hinted_locations.add(loc)
+    for area_name, locations in locations_per_area.items():
+        not_yet_hinted_locations = sum(location not in already_hinted_locations for location in locations)
+        unhinted_locations_per_area[area_name] = {loc for loc in locations if loc not in already_hinted_locations}
+        unhinted_location_percentage_per_area[area_name] = not_yet_hinted_locations / len(locations)
 
-    next_random_hint_is_item = world.random.randrange(0, 2)
+    items_per_area = {area_name: [location.item for location in locations]
+                      for area_name, locations in locations_per_area.items()}
 
-    while len(hints) < hint_amount:
-        if next_random_hint_is_item:
-            if not prog_items_in_this_world:
-                next_random_hint_is_item = not next_random_hint_is_item
-                continue
+    areas = sorted(area for area in items_per_area if unhinted_location_percentage_per_area[area])
+    weights = [unhinted_location_percentage_per_area[area] for area in areas]
 
-            hint = make_hint_from_item(world, prog_items_in_this_world.pop(), own_itempool)
+    amount = min(amount, len(weights))
+
+    hinted_areas = weighted_sample(world.random, areas, weights, amount)
+
+    return hinted_areas, unhinted_locations_per_area
+
+
+def get_hintable_areas(world: "WitnessWorld") -> Tuple[Dict[str, List[Location]], Dict[str, List[Item]]]:
+    potential_areas = list(StaticWitnessLogic.ALL_AREAS_BY_NAME.keys())
+
+    locations_per_area = dict()
+    items_per_area = dict()
+
+    for area in potential_areas:
+        regions = [
+            world.regio.created_regions[region]
+            for region in StaticWitnessLogic.ALL_AREAS_BY_NAME[area]["regions"]
+            if region in world.regio.created_regions
+        ]
+        locations = [location for region in regions for location in region.get_locations() if location.address]
+
+        if locations:
+            locations_per_area[area] = locations
+            items_per_area[area] = [location.item for location in locations]
+
+    return locations_per_area, items_per_area
 
-            if not hint or hint[0] in already_hinted_locations:
-                continue
 
-            hints.append((f"{hint[1]} can be found at {hint[0]}.", hint[2]))
+def word_area_hint(world: "WitnessWorld", hinted_area: str, corresponding_items: List[Item]) -> str:
+    """
+    Word the hint for an area using natural sounding language.
+    This takes into account how much progression there is, how much of it is local/non-local, and whether there are
+    any local lasers to be found in this area.
+    """
 
-            already_hinted_locations.add(hint[0])
+    local_progression = sum(
+        item.player == world.player
+        and item.classification in {ItemClassification.progression, ItemClassification.progression_skip_balancing}
```

Comment:
I believe these are equivalent
```suggestion
        and item.advancement
```